### ISSUE TITLE
Parallelism of proof processing in async mode?

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#15998]
These experiments with async mode give unexpected results.  I'm running them in CoqIDE, 8.15.1+, Ubuntu on Windows/WSL2:

1. Running `Pff.v` from `ci-flocq` takes 1:37 both with async enabled (i.e., on by default) and without async (I passed `-async-proofs off`).  I only see one worker active in the Jobs panel.  (Both runs included the changes in https://github.com/coq/coq/pull/15996, which are irrelevant to the comparison.)

2.  The following loops are tuned to print a message every second or so.  There doesn't appear to be any interleaving. I only see only one active worker in the Jobs panel. I expected to see interleaving of the output and 2 active jobs.

```
Lemma D: True. Proof. idtac.  exact I.  Qed.
Lemma B: True. Proof. do 10 (idtac "B"; do 2000000 idtac). idtac "done B". exact I. Qed.
Lemma C: True. Proof. do 10 (idtac "C"; do 2000000 idtac). idtac "done C". exact I. Qed.
Lemma A: True. Proof. idtac.  exact I.  Qed.
```

Output:

```
B
B
B
B
B
B
B
B
B
B
done B
C
C
C
C
C
C
C
C
C
C
done C
```

Attn : @gares

**Comment by gares** [comment]
did you try -async-proofs-j 2?
there are many reasons why coq may decide to not delegate proofs, like open sections and no proof using. I don't know pfff enough to tell. Enabling stm debugging should print some info

**Comment by jfehrle** [comment]
> did you try -async-proofs-j 2?

I thought that was the default on Linux.  I'll try that in the morning.

How about the second example?  Any other reason it would it not run in parallel?

**Comment by jfehrle** [comment]
And I'll review the description in the doc of limitations for async.

**Comment by gares** [comment]
The (sane) default is 1, because of the memory footprint of each worker.
You have 1 worker crunching on async proofs, and 1 master answering to CoqIDE and crunching on non-async sentences

**Comment by jfehrle** [comment]
OK, using `-async-proofs-j 2` makes the second example work as expected, although if I reset the buffer and run the script a second time, there are 2 idle workers that hang around until the script completes.  Should I file a bug report?

![image](https://user-images.githubusercontent.com/1253341/167469579-9885e3a0-ce0a-426d-842e-84514a5002a1.png)

> You have 1 worker crunching on async proofs, and 1 master answering to CoqIDE and crunching on non-async sentences. This already doubles the memory footprint.

"Doubling the memory footprint" deserves some unpacking.  When you fork a process, the memory pages of the parent and child should be shared until the first write into the page by either process, at which time the writing process gets an unshared copy.  There could be a high degree of sharing, however shared pages of the heap that still have free space are likely to soon be allocated and thus become unshared.  A heap compaction before the fork may help although it is itself an expensive operation.  Heap space for transient objects specific to a proof can't be shared.

Or perhaps it would be possible to note the amount of incremental memory used for each proof and use that for smarter scheduling of the next run, varying the number of active workers?  I suspect many proofs only need a modest amount of memory.  It would be interesting to get statistics to support or contradict that suspicion, particularly if you're interested in improving the parallel mechanism.

My impression is that faster builds are very high on many users' wish lists.  If so, that makes it an interesting thing for me to (eventually) work on.

I got no speedup using 2 workers on Pff.v, which has 28K lines, 71 sections and no `Proof using` annotations.  I suspect almost everything is in some section.  Is that sufficient to disable parallelism?  Is it technically feasible to add some (fast) overall analysis of a script so that the `Proof using` annotations are not needed?

Is there any plan to use eventually use threading instead of forking processes, which may use somewhat less memory?

**Comment by gares** [comment]
It is normal, if the delegated task turns out to be too quick it is not delegated again.

We don't fork, not even on unix. This changes in the new LSP thingy, but the STM is like that, it works the same on windows and unix.

Async proofs are not to give a speedup. They might, but their reason of existence is to be able to jump forward in a file and have coq catch up in the background. It is about having a more responsive system.

One cannot predict without running, but once you run you know what to put in Proof using, see Set Suggest Proof Using.

No, threading in OCaml gives no parallelism, up to version 5. Also, Coq has so much messy global state, that it is hopeless to use threads.

**Comment by jfehrle** [comment]
I meant threading when Coq is on OCaml 5.  Not sure that it's hopeless, but probably not the first or second thing I'd try, either.

You were not too concerned about performance per se.  Does the following seem correct to you?  Or no opinion?

> My impression is that faster builds are very high on many users' wish lists. If so, that makes it an interesting thing for me to (eventually) work on.

**Comment by gares** [comment]
It depends what you mean by build. For example building .vos does not build proofs, but you xan use all the rest and go back to work while, in the background, you also check proofs.

Maybe a parallel with more standard compilers is building fast vs building optimized code.
My objective was to make the edit/build/play/edit loop faster. For that you don't need the same notion of build.

It is not very different in spirit that rewriting your java compiler so that it gives you feedback soon in the editor.

**Comment by jfehrle** [comment]
IIRC, @Blaisorblade said that recompiling their proofs took many hours, which was burdensome.  If there are reasonable ways to increase parallelism that would benefit both compiling and interactive use.  Worth attempting, yes?  An in interesting challenge, too.

**Comment by mrhaandi** [comment]
> If there are reasonable ways to increase parallelism that would benefit both compiling and interactive use it would be worth attempting, right? And an interesting challenge, too.

I strongly agree. The 150K LOC [coq-library-undecidability](https://github.com/uds-psl/coq-library-undecidability) would definitely benefit from this. Almost every script sets `Default Proof Using "Type"`. 

I tried `time coqc -async-proofs on -async-proofs-j x` with `x = 1..4` for a 1K LOC file with curious results :
- `-async-proofs-j 1` : `50.899s` (1x CPU load)
- `-async-proofs-j 2` : `33.921s` (~2x CPU load)
- `-async-proofs-j 3` : `32.440s` (~2.5x CPU load)
- `-async-proofs-j 4` : `30.998s` (~3x CPU load)
- `-async-proofs off` : `10.447s` (1x CPU load)

Overall, `-async-proofs-j x` with `x > 1` improves compilation time, but setting `-async-proofs on` is too expensive in the first place.

Of course, as @gares says
> Async proofs are not to give a speedup. They might, but their reason of existence is to be able to jump forward in a file and have coq catch up in the background. It is about having a more responsive system.

It would be nice to have both a responsive system and parallel proof-level compilation, as parallel file-level compilation does not work for linear dependency chains.

**Comment by gares** [comment]
Do I read it right that the file takes 10 seconds to compile? That is a very quick file for the current level of parallelism, I'm afraid.

For the whole library you may still try vio or vos, that should give you more reposive "batch compilation", even if your files are quick. Can you share the timing of `coqc -vos` and `coqc -vio` for this file `x`?

**Comment by gares** [comment]
Ah great. And `coqc -schedule-vio-checking 4 x.vio`?

**Comment by gares** [comment]
(that would be the the best we get if reduce marshaling time / worker spawning time to zero)

**Comment by mrhaandi** [comment]
> Ah great. And `coqc -schedule-vio-checking 4 x.vio`?

- `coqc -schedule-vio-checking 4 x.vio` :  `5.155s`
- `coqc -schedule-vio-checking 6 x.vio` :  `3.985s` (6 is my CPU count)

**Comment by gares** [comment]
If this file is representative then:
- with -vos you can get back to work after a change 20x faster
- with -vio you can get back to work after a change 15x faster, but
  - adding -schedule-vio-checking you can get a good confidence level check of your library 2x faster

So I'm not so sure what we try to obtain with `coqc -async-proofs on`. Yes, you get the same level of confidence that you have with `coqc`, which you don't get with vok or vio2vo. Is there any other motivation?

Anyway, many thanks for sharing your data and for setting things up so that vos/vio do work on your library. I'll keep that in mind next time I have to run benchmarks.

**Comment by ppedrot** [comment]
> Yes, you get the same level of confidence that you have with coqc, which you don't get with vok or vio2vo.

That's not even true, isn't it? There is still the problem that global universe constraints might be computed in the wrong order. I think this is pretty much the same confidence as quick + vio2vo.

**Comment by gares** [comment]
Since when the order of constraints affects consistency? Anyway, running a good old plain `coqc` and even `coqchk` in CI is recommended.

**Comment by mrhaandi** [comment]
> Is there any other motivation?

We quite effectively use `-vos` as you mentioned, and mostly it works well.
Still, we regularly have universe inconsistencies and sometimes wrong `Proof using` annotations which get unnoticed at first, and then break upon full compilation.
Usually, this is only a slight annoyance. Sometimes it makes me sad, that despite all annotation effort the final compilation on a 16 core system hangs forever on 10% CPU utilization in a linear proof chain.

**Comment by ppedrot** [comment]
> Since when the order of constraints affects consistency?

It doesn't directly, but it affects the output, which qualifies as Pollack inconsistency. Here is a contrived but small example of this phenomenon.
```
Universe u v.

Lemma foo : unit.
Proof.
refine (let _ : Type@{u} := Type@{v} in tt).
Qed.

Definition magic_bool : bool.
Proof.
(let _ := constr:(Type@{u} : Type@{v}) in exact true) || exact false.
Defined.

Eval compute in magic_bool.
(* false in sequential mode, true with async *)
```

**Comment by mrhaandi** [comment]
> Is there any other motivation?

Another motivation is when working in the editor `-async-proofs on` is a must for a smooth workflow. (And already solves the largest issue.)
The mentioned `50s` or `30s` with `-async-proofs-j 2` (instead of `4s` as suggested by the ideal case) are the exact times to check the (possibly incomplete) file in the editor. Any speedup here is, of course, appreciated.

**Comment by jfehrle** [comment]
I've hardly started looking at this.  I noticed poor scaling as you have.  A few thoughts:
- For very large files (e.g. Pff.v in flocq, 1.1MB), proof workers seem to take precedence over passing statements to the server, limiting the number of proofs available for processing
- I'd like to measure how long it takes to start a worker.  Perhaps using fork instead of exec would reduce overhead.

IINM, the async mechanism should generally make it possible to do most of the proofs in parallel, which would make it promising for highly parallel processing.  Reducing the per-proof overhead will always help.  And if that expectation is not right, it would be worthwhile to quantify the potential for parallelism.

> @mrhaandi despite all annotation effort the final compilation on a 16 core system hangs forever on 10% CPU utilization in a linear proof chain.

The "linear proof chain" is within a single script or dependencies across scripts?  The former may be amenable to simple improvements, the latter may be require significant work.  Which case are you more concerned about?

**Comment by mrhaandi** [comment]
> The "linear proof chain" is within a single script or dependencies across scripts? The former may be amenable to simple improvements, the latter may be require significant work. Which case are you more concerned about?

A linear chain of individual proof scripts (files, which are essentially large sections and induce artificial overhead upon factorization). So `A.v` requires `B.v` which requires `C.v`...
Here, file-level parallelism does not help and there is no automatic proof-level parallelism.

**Comment by jfehrle** [comment]
@gares For Pff.v in async mode, every request to the proof worker is about 13MWords because it includes `VCS.slice ~block_start:t_start ~block_stop:t_stop`.  The difference between the t_start and t_stop stateids is at most 11.  Is all of that data really necessary?  Could there be easy ways to cut that way back?

I see every proof taking at least .2 second (from request sent to receiving `Response`.  That seems slow.

**Comment by gares** [comment]
The data sent is a large over approximation. But as I was saying somewhere else, the new code I'm working is based on fork (on unix) so I don't want to invest my time on this.

**Comment by jfehrle** [comment]
I didn't know you were working on new code.  How much code are you planning to replace? Is that the Stm replacement?  Is it worthwhile for me to poke around in this now?

**Comment by gares** [comment]
Yes, it is the STM replacement, the new document manager. It shall be "ready" in September. *But* it will all be about interactive mode, and we are likely to have engineers working on the UI part (vscode) from then.

IMO it was a mistake to not separate batch and interactive use. Interactive use alone is simpler, e.g. one does not really need the workers to send back proofs and somehow build a vo, and reactivity is more important than completion time. Similarly batch mode alone has much more to gain from separating the compilation into phases (today you gain 20x on a single core if you look at round trip time, I don't even have 20 cores...). Sure, some code is shared, and some internal problems are also, but as we see here trying to use the interactive part in batch mode does not just work, and I'm afraid it is too hard to make it 10 times faster.

Last, if we want to parallelize more, I think we should first work on the surface language, a bit like we did for `par:`. For example a long proof can surely be split into independent chunks, but delimiters like `{` and `}` (or bullets) don't guarantee as of today that these parts are actually independent. If we had that hint from the user, we could consider delegating these non-atomic sub proofs to a pool of workers, as `par` does for atomic-and-closing proof steps. I personally don't believe that applying "brute force / blind" parallelism is both doable (given the source code/architecture of Coq) and profitable.

**Comment by Blaisorblade** [comment]
https://github.com/coq/coq/issues/15998#issuecomment-1144648062

> We quite effectively use -vos as you mentioned, and mostly it works well.
Still, we regularly have universe inconsistencies and sometimes wrong Proof using annotations which get unnoticed at first, and then break upon full compilation.

Shouldn't "wrong Proof using" get noticed in `vok` mode?

**Comment by mrhaandi** [comment]
> Shouldn't "wrong Proof using" get noticed in `vok` mode?

Yes. However, batch compilation needs to happen anyway to be on the safe side. Additionally running `vok` to notice problems slightly earlier is not always worth the additional CPU cycles (specifically for our project).

**Comment by jfehrle** [comment]
@gares The new Stm will be fully or largely compatible with the current XML protocol?  Is there a description of any new parts that require IDE support?

> For example a long proof can surely be split into independent chunks, but delimiters like { and } (or bullets) don't guarantee as of today that these parts are actually independent.

Perhaps we should encourage developers to split things up themselves, especially if the text of the theorem is long (as opposed to the run time)?  Just like splitting up large functions is good practice in programming languages.  Is there any reason that would be difficult for users?

I expect to poke around in this some more and understand the current performance better.  I'll let you know if I find anything that looks promising.

> I personally don't believe that applying "brute force / blind" parallelism is both doable (given the source code/architecture of Coq) and profitable.

Do you have any statistics or instrumentation that tell you the degree of parallelism that's possible for a script?

**Comment by Blaisorblade** [comment]
> Perhaps we should encourage developers to split things up themselves, especially if the text of the theorem is long (as opposed to the run time)? Just like splitting up large functions is good practice in programming languages. Is there any reason that would be difficult for users?

For one, text length and runtime of proof scripts seem very weakly correlated in general.

But more importantly, people should factor functions using sensible software engineering, and that suggests different ways of splitting theorems. I have an example below (tho it's very abstract, sorry).

## Before
 Consider splitting out the proof done by `part_B` in this proof script:
```coq
Theorem name : statement.
Proof.
  part_A. 
  { (* Here the goal is statement_B. *) part_B. }
  part_C.
Qed.
```

Here, `statement_B` is generated code that's not written down anywhere — it might change during project evolution, but the proof will often keep working. With something like `par:` or similar, we could run `part_B` in parallel without other changes.

## After

Lifting:

```coq
Lemma aux_B : statement_B.
Proof. part_B. Qed.

Theorem name : statement.
Proof.
  part_A. 
  { apply aux_B. }
  part_C.
Qed.
```

Now, `statement_B` is part of the source code, and might break if we change `part_A`, or `statement`, or things that `statement` refers to. If `part_A` uses `unfold foo`, as it is common, then `statement_B` will depend on `foo`'s body. Not great for maintenance.

TLDR: this refactoring makes little sense. There are multiple approaches to refactor things instead, but they won't solve all build problems, so commands to split subproofs make sense.

**Comment by jfehrle** [comment]
@Blaisorblade Yeah, if `statement_B` can change, extracting it as a lemma would create a maintenance issue, so adding/improving language constructs to ensure that the enclosed chunk can run as a separate task makes sense.

OTOH, how does a proof developer know where to use these constructs?  I think they would need instrumentation to learn where to add constructs such as `par:`, such as a convenient way to see the processing time required for each tactic.  The UI could show a table with the hot spots sorted by execution time.  Clicking on an entry would take you to the relevant part of the script.  The slowest parts of the script could be highlighted.  (My impression is that our current perf tools are much more granular and have no UI support.)

That would perhaps enable
1. adding `par:` where helpful,
2. the developer substituting alternate, faster tactics and/or
3. identifying Coq performance bugs with specific tactics

though I don't know how realistic # 2 and # 3 are.  (WDYT?)

Perhaps it would be interesting to show the derivation graph of the goals showing the processing time for each subgraph as a table or tree with the ability to click on a subgraph to highlight it in the script.  More complex to implement, not sure it adds that much over the simpler idea above.

**Comment by SkySkimmer** [comment]
Maybe we could make some kind of `Set Bullet Behavior "independent"` mode within which separate bullets or braces can be run in parallel. Probably most proofs would fit.

**Comment by jfehrle** [comment]
A tricky bit is telling whether making the part in bullets/braces a parallel task will make the proof complete faster.  There's a cost to create parallel tasks.  That will slow down the the (probably) single process doing the splitting.
And if there's no core free to work on the task, splitting could make you run slower.  If you can get the overhead low enough, it may not matter.

**Comment by mrhaandi** [comment]
This discourse discussion could also be of relevance: https://coq.discourse.group/t/ad-hoc-asynchronicity/1079

**Comment by jfehrle** [comment]
I just found something unexpected: `Theorem`s and `Lemma`s that don't contain "Proof." are not delegated to workers.  As a matter of fact, the file I've been playing with, `Pff.v`, has 730 theorems and lemmas and only 73 `Proof`s.  So no wonder I see no speed up with more processes.

@gares Did you know about this?  How hard would it be to fix this?  If you tell me what to change, maybe I could do it.

**Comment by gares** [comment]
Yes. From https://coq.inria.fr/refman/addendum/parallel-proof-processing.html#proof-annotations

> The presence of Proof using is needed to process proofs asynchronously in interactive mode.

I think that should find the time to look at that chapter of the refman.

**Comment by jfehrle** [comment]
@gares @Blaisorblade @mrhaandi Perhaps you or others would comment on #16151?

**Comment by jfehrle** [comment]
@gares @Blaisorblade @mrhaandi Do you agree with these [comments above](https://github.com/coq/coq/issues/15998#issuecomment-1146863089) about the need for tooling?  Would they be useful today or would they only become useful with new language features to support async?  (I wouldn't spend time on that without some confirmation from users such as you all.)

**Comment by mrhaandi** [comment]
> Perhaps it would be interesting to show the derivation graph of the goals showing the processing time for each subgraph as a table or tree with the ability to click on a subgraph to highlight it in the script.

I'm not entirely sure about individual goals, but what I sometimes need to do by hand is measure compilation time of a lemma and compose a lemma dependency tree. After all, the total compilation time on a highly parallel architecture is that of the lemma dependency chain with with most total runtime. This allows for refactoring into sensible `.v` files with most file-level parallelism.
Command-line tooling for this would be great. Refactoring individual goals feels less impactful.

**Comment by jfehrle** [comment]
> I'm not entirely sure about individual goals

Well, how would you figure out which parts of a proof would benefit from `async` (as in your proposal) or `par:`?  You want to identify the hot spots and tweak them.

**Comment by mrhaandi** [comment]
> Well, how would you figure out which parts of a proof would benefit from `async` (as in your proposal) or `par:`? You want to identify the hot spots and tweak them.

What I mean is, if your question is about useful tooling in general, then `par:` is of less concern to me compared to file-level parallelism tools. If your question is about tooling wrt. to `par: / async` then I'm not sure because the `par:` feature is cumbersome and general `async` does not exist.

**Comment by jfehrle** [comment]
> @gares there are many reasons why coq may decide to not delegate proofs, like open sections and no proof using.  I don't know pfff enough to tell. Enabling stm debugging should print some info

After adding `Proof using` everywhere (635 places), I still only see 235 tasks delegated to workers.  How do I enable stm debugging?  Maybe we could describe the reasons for not delegating and, if possible, how to get around them by script changes.

This fails (IDE shows coqidetop died dialog):
```
$ time dune exec -- dev/shim/coqide-prelude -async-proofs-j 4 -d proof-using
Don't know what to do with proofworker:0
```

`Test Debug` says:

![image](https://user-images.githubusercontent.com/1253341/173201335-c656a3a4-b06f-4fb5-81f5-69081764c341.png)

**Comment by jfehrle** [comment]
> @gares It is normal, if the delegated task turns out to be too quick it is not delegated again.

Is this information saved only within the coqtopide process and not in any file?  So if you start a fresh copy of coqtopide it will always delegate?

**Comment by gares** [comment]
within the same run, if you jump back and forth. or in the aux file.

**Comment by jfehrle** [comment]
> @ppedrot There is still the problem that global universe constraints might be computed in the wrong order. I think this is pretty much the same confidence as quick + vio2vo.

Would it be worthwhile to create a new annotation to give the correct order?  With messages akin to `Suggest Proof Using`.

**Comment by Alizter** [comment]
@jfehrle I don't believe that is feasible since the universe constraints are modified when each subprocess finishes. If we have a file which spawns several workers, which ever finishes first gets to modify the universe constraints. If we force them to follow a specific order, then we have to delay the processing of the proofs until they have their turn, which won't be much different than not spawning workers at all.

This is at least my understanding of the situation.

**Comment by jfehrle** [comment]
If the universe constrains are given as an annotation on the theorem, is it really that hard to get the workers to be consistent with the annotation rather than modifying the universe constants?  Presumably we create tooling to suggest or add the correct annotation.

**Comment by ppedrot** [comment]
We need two things:
- A way to have private monomorphic constraints inside Qed blocks 
- A warning when a Qed block creates non-private constraints (+ the corresponding attribute to disable it).

Proofs that create non-private constraints are somewhat "dangerous" to defer because the result will depend on the scheduler. The warning would allow annotating these, but without private constraints it would fire 100% of the time.

**Comment by ejgallego** [comment]
@ppedrot you only need that if your document manager client is pretty dumb and doesn't know about incremental computation and caching. Indeed, once you have a proper client, 1 is just an optimization. 2 can be replaced by a notification mechanism, which tell the client how the "proof hole" was filled, for example, if the filling affected the constraint graph, so the client triggers rebuild of deps then. The tricky part is doing this efficienctly.

**Comment by jfehrle** [comment]
Wouldn't you also have the scheduling issue running `coqc -async-proofs on`?  In which case there wouldn't be a document manager client.

**Comment by ppedrot** [comment]
1 is not an optimization, it's literally a change of semantics.

**Comment by ejgallego** [comment]
@ppedrot indeed, I mean an optimization in the sense of that it allows for an intra-document dep graph with better shape.

@jfehrle `coqc -async-proofs on` uses a document manager to precisely understand (and handle) proof structure.

**Comment by jfehrle** [comment]
@ejgallego I took "document manager client" to mean code in the IDE.  On general principles that should be part of the Coq server (coqidetop) rather than implemented repeatedly in each IDE.

**Comment by jfehrle** [comment]
Any comments on https://github.com/coq/coq/issues/16266?

**Comment by jfehrle** [comment]
Also, any comments on [#16273 Support async processing when compiling multiple dependent files](https://github.com/coq/coq/issues/16273)?
